File: src/incode_mcp/acceptance.py
"""Pure correctness metrics shared by accelerator promotion gates."""

from __future__ import annotations

from typing import Any

import numpy as np
from numpy.typing import NDArray

FloatMatrix = NDArray[np.float32]


def _matrix(values: NDArray[Any], *, name: str) -> FloatMatrix:
    matrix = np.asarray(values, dtype=np.float32)
    if matrix.ndim != 2 or not matrix.shape[0] or not matrix.shape[1]:
        raise ValueError(f"{name} must be a non-empty two-dimensional matrix")
    if not np.all(np.isfinite(matrix)):
        raise ValueError(f"{name} contains non-finite values")
    return matrix


def _normalized(values: NDArray[Any], *, name: str) -> FloatMatrix:
    matrix = _matrix(values, name=name)
    norms = np.linalg.norm(matrix, axis=1, keepdims=True)
    if np.any(norms <= np.float32(1e-12)):
        raise ValueError(f"{name} contains a zero-length row")
    return np.asarray(matrix / norms, dtype=np.float32)


def cosine_rows(reference: NDArray[Any], candidate: NDArray[Any]) -> FloatMatrix:
    """Return cosine similarity for each corresponding pair of matrix rows."""

    left = _normalized(reference, name="reference")
    right = _normalized(candidate, name="candidate")
    if left.shape != right.shape:
        raise ValueError(f"reference and candidate shapes differ: {left.shape} != {right.shape}")
    return np.asarray(np.sum(left * right, axis=1), dtype=np.float32)


def top_k_overlap(
    queries: NDArray[Any],
    reference: NDArray[Any],
    candidate: NDArray[Any],
    *,
    k: int,
) -> float:
    """Return mean top-k result overlap for two candidate vector matrices."""

    query_rows = _normalized(queries, name="queries")
    reference_rows = _normalized(reference, name="reference")
    candidate_rows = _normalized(candidate, name="candidate")
    if reference_rows.shape != candidate_rows.shape:
        raise ValueError(
            f"reference and candidate shapes differ: "
            f"{reference_rows.shape} != {candidate_rows.shape}"
        )
    if query_rows.shape[1] != reference_rows.shape[1]:
        raise ValueError(
            f"query dimension {query_rows.shape[1]} does not match "
            f"candidate dimension {reference_rows.shape[1]}"
        )
    if not 1 <= k <= reference_rows.shape[0]:
        raise ValueError(f"k must be from 1 to {reference_rows.shape[0]}")

    reference_order = np.argsort(
        -(query_rows @ reference_rows.T),
        axis=1,
        kind="stable",
    )[:, :k]
    candidate_order = np.argsort(
        -(query_rows @ candidate_rows.T),
        axis=1,
        kind="stable",
    )[:, :k]
    overlap = sum(
        len(set(left.tolist()).intersection(right.tolist()))
        for left, right in zip(reference_order, candidate_order, strict=True)
    )
    return float(overlap / (query_rows.shape[0] * k))

File: ts/packages/spikes/src/acceptance.ts
/**
 * Port of `acceptance.py` -- the pure correctness metrics the accelerator
 * promotion gates already use.
 *
 * S3 scores the TypeScript embedder against Python-produced vectors with the
 * *same* metrics the project already trusts, rather than inventing a parity
 * measure for the migration. Its numbers are directly comparable to what
 * `test_accelerator_acceptance.py` asserts.
 *
 * **The canonical port now lives in `packages/server/src/acceptance.ts`** --
 * Phase 1 landed it, with `topKRankCorrelation` and a proper suite. This copy
 * stays deliberately: a Phase 0 spike is a record of an experiment, and one that
 * reached back into code written after it ran would no longer reproduce the
 * verdict in `docs/plans/2026-08-17-phase-0-spike-results.md`. Fix bugs in the
 * server module; change this one only to keep the spike runnable.
 *
 * Deliberately written over plain arrays: pulling in a linear-algebra
 * dependency to answer a Phase 0 question would prejudge a Phase 4 decision.
 */

export type Matrix = readonly (readonly number[])[];

function validate(matrix: Matrix, name: string): void {
  const rows = matrix.length;
  const width = matrix[0]?.length ?? 0;
  if (rows === 0 || width === 0) {
    throw new Error(`${name} must be a non-empty two-dimensional matrix`);
  }
  for (const row of matrix) {
    if (row.length !== width) throw new Error(`${name} is ragged`);
    for (const value of row) {
      if (!Number.isFinite(value)) throw new Error(`${name} contains non-finite values`);
    }
  }
}

/** L2-normalize each row, rejecting zero-length rows as `acceptance.py` does. */
export function normalized(matrix: Matrix, name: string): number[][] {
  validate(matrix, name);
  return matrix.map((row) => {
    const norm = Math.sqrt(row.reduce((total, value) => total + value * value, 0));
    if (norm <= 1e-12) throw new Error(`${name} contains a zero-length row`);
    return row.map((value) => value / norm);
  });
}

/** Cosine similarity for each corresponding pair of matrix rows. */
export function cosineRows(reference: Matrix, candidate: Matrix): number[] {
  const left = normalized(reference, "reference");
  const right = normalized(candidate, "candidate");
  if (left.length !== right.length || left[0]?.length !== right[0]?.length) {
    throw new Error(
      `reference and candidate shapes differ: ` +
        `${left.length}x${left[0]?.length} != ${right.length}x${right[0]?.length}`,
    );
  }
  return left.map((row, index) => {
    const other = right[index];
    if (other === undefined) throw new Error("row count mismatch");
    return row.reduce((total, value, column) => total + value * (other[column] ?? 0), 0);
  });
}

/** Indices of the k highest scores, ties broken by ascending index (stable). */
function topK(scores: readonly number[], k: number): number[] {
  return scores
    .map((score, index) => ({ score, index }))
    .sort((left, right) => right.score - left.score || left.index - right.index)
    .slice(0, k)
    .map((entry) => entry.index);
}

/** Mean top-k result overlap between two candidate vector matrices. */
export function topKOverlap(
  queries: Matrix,
  reference: Matrix,
  candidate: Matrix,
  k: number,
): number {
  const queryRows = normalized(queries, "queries");
  const referenceRows = normalized(reference, "reference");
  const candidateRows = normalized(candidate, "candidate");

  if (referenceRows.length !== candidateRows.length) {
    throw new Error(
      `reference and candidate shapes differ: ${referenceRows.length} != ${candidateRows.length}`,
    );
  }
  if (queryRows[0]?.length !== referenceRows[0]?.length) {
    throw new Error(
      `query dimension ${queryRows[0]?.length} does not match ` +
        `candidate dimension ${referenceRows[0]?.length}`,
    );
  }
  if (k < 1 || k > referenceRows.length) {
    throw new Error(`k must be from 1 to ${referenceRows.length}`);
  }

  const score = (query: readonly number[], rows: readonly (readonly number[])[]): number[] =>
    rows.map((row) =>
      row.reduce((total, value, column) => total + value * (query[column] ?? 0), 0),
    );

  let overlap = 0;
  for (const query of queryRows) {
    const referenceOrder = new Set(topK(score(query, referenceRows), k));
    for (const index of topK(score(query, candidateRows), k)) {
      if (referenceOrder.has(index)) overlap += 1;
    }
  }
  return overlap / (queryRows.length * k);
}

/**
 * Attention-mask mean pooling followed by L2 normalization.
 *
 * `direct_onnx.py::mean_pool_and_normalize`, which is the pooling the index
 * was built with. Getting this wrong produces vectors that look plausible and
 * rank differently, which is exactly the failure S3 exists to catch.
 */
export function meanPoolAndNormalize(
  hidden: Float32Array,
  attentionMask: readonly (readonly number[])[],
  shape: readonly [number, number, number],
): number[][] {
  const [batch, sequence, width] = shape;
  const pooled: number[][] = [];
  for (let row = 0; row < batch; row += 1) {
    const mask = attentionMask[row];
    if (mask === undefined) throw new Error(`attention mask is missing row ${row}`);
    const sums = new Array<number>(width).fill(0);
    let counted = 0;
    for (let token = 0; token < sequence; token += 1) {
      if ((mask[token] ?? 0) === 0) continue;
      counted += 1;
      const base = (row * sequence + token) * width;
      for (let column = 0; column < width; column += 1) {
        sums[column] = (sums[column] ?? 0) + (hidden[base + column] ?? 0);
      }
    }
    // numpy divides by the clamped mask sum; a fully-masked row would be a bug
    // upstream, so match the clamp rather than inventing a zero vector.
    const divisor = Math.max(counted, 1e-9);
    const averaged = sums.map((value) => value / divisor);
    const norm = Math.sqrt(averaged.reduce((total, value) => total + value * value, 0));
    pooled.push(averaged.map((value) => value / Math.max(norm, 1e-12)));
  }
  return pooled;
}
